crates/provider/src/layers/batch.rs

```diff
@@ -311,18 +342,45 @@ impl<P: Provider<N> + 'static, N: Network> CallBatchBackend<P, N> {
         }
     }
 
-    fn process_msg(&mut self, msg: CallBatchMsg) {
+    fn process_msg(&mut self, msg: CallBatchMsg<N>) {
         self.pending.push(msg);
     }
 
     async fn send_batch(&mut self) {
-        let result = self.send_batch_inner().await;
         let pending = std::mem::take(&mut self.pending);
+
+        if pending.len() == 1 {
+            let msg = pending.into_iter().next().unwrap();
+
+            let result: Result<CallResult, _> = match msg.kind {
+                CallBatchMsgKind::Call(tx) => {
+                    let res = self.inner.call(tx).await.unwrap();
+                    Ok(CallResult::Single(res))
+                }
+                CallBatchMsgKind::BlockNumber => {
+                    let result = self.inner.get_block_number().into_future().await.unwrap();
+                    Ok(CallResult::Single(Bytes::from(result.to_be_bytes())))
+                }
+                CallBatchMsgKind::ChainId => {
+                    let result = self.inner.get_chain_id().into_future().await.unwrap();
+                    Ok(CallResult::Single(Bytes::from(result.to_be_bytes())))
+                }
+                CallBatchMsgKind::Balance(addr) => {
+                    let result = self.inner.get_balance(addr).into_future().await.unwrap();
+                    Ok(CallResult::Single(Bytes::from(result.to_be_bytes::<32>())))
```

Comment:
these `to_be_bytes` happen to work fine but we want to make sure they're ABI encoded with `.abi_encode()` instead